Haskell Parser Combinators/article.md

```diff
@@ -0,0 +1,1330 @@
+
+In other words, make sure that you parse a boolean in `atom` before you parse an identifier.
+
+### S-expressions
+
+Finally, let's move on to our last parser, which parses S-expressions themselves, such as `(foo 42 "hey")` or `(bar)`.
+
+First, we will do it the naïve way, which may come as the first natural solution to someone just learning Megaparsec, and then we will do it in a better manner using extra tools offered by Megaparsec.
+
+#### S-expressions: The awkward way
+
+The idea here is to parse an atom, followed by 0 or more atoms. Of course, this means that we want to use `many` to parse such atoms. Additionally, an S-expression may be surrounded by parenthesis, so we can use `between` for such a task.
+
+This parser, however, is not as simple as before, but let's first do it the wrong way (which won't work properly), analyze the problem, and come up with a solution to it.
+
+```hs
+sexp :: Parser (SExp, [SExp])
+sexp = label "S-expression" $ between (char '(') (char ')') ((,) <$> atom <*> many atom)
+-- Or:
+-- sexp = label "S-expression" $ char '(' *> ((,) <$> atom <*> many atom) <* char ')'
+```
+
+After adding it to `atom` with `uncurry SSExp <$> sexp`, we can test it in GHCi:
+
+```hs
+>>> parseTest atom "(foo)"
+SSExp (SId (Identifier {getId = "foo"})) []
+```
+
+Ok, looks good so far. What if we have arguments being applied to it?
+
+```
+>>> parseTest atom "(foo bar)"
+1:5:
+  |
+1 | (foo bar)
+  |     ^
+unexpected space
+expecting ')', '_', S-expression, alphanumeric character, boolean, identifier, integer, or string
+```
+
+Uh-oh. We can't parse spaces, as the error makes it evident. Either we close parenthesis (stop parsing this S-expression), continue the definition of the identifier (`_` or `alphanumeric character`), put another atom (`boolean, identifier, integer, or string`), or another S-expression (by opening parenthesis).
+
+One solution that could come to one's mind is to use the function `sepBy`, which takes a parser and a separator, and use it as ``atom `sepBy` space``. In this case, `space` is a parser defined by Megaparsec that skips zero or more spaces.
+
+```hs
+>>> parseTest (between (char '(') (char ')') $ SSExp <$> atom <*> (space *> atom `sepBy` space)) "(foo bar)"
+SSExp (SId (Identifier {getId = "foo"})) [SId (Identifier {getId = "bar"})]
+```
+
+Of course, this may raise other questions: what if we have a space character before or after the opening and closing parenthesis?
+
+While that works, the [`megaparsec` documentation](https://hackage.haskell.org/package/megaparsec-9.2.0/docs/Text-Megaparsec-Char-Lexer.html) already offers a solution for this kind of problem, namely, a lexer. Indeed, we will see that some of our previous definitions could have been written more elegantly with the help of this module.
```

Comment:
That works = having a space character? Perhaps some rephrasing is necessary.

Reply:
I meant explicitly skipping spaces. Reworded it.